File: py_src/paperjam/_stamp.py
"""Page stamping/overlay for Document."""

from __future__ import annotations

from paperjam import _paperjam
from paperjam._document import Document


def _stamp(
    self: Document,
    stamp_doc: Document,
    *,
    source_page: int = 1,
    target_pages: list[int] | None = None,
    x: float = 0.0,
    y: float = 0.0,
    scale: float = 1.0,
    opacity: float = 1.0,
    layer: str = "over",
) -> Document:
    """Overlay a page from another PDF onto pages of this document."""
    inner = self._ensure_open()
    stamp_inner = stamp_doc._ensure_open()
    new_inner = _paperjam.stamp_pages(
        inner,
        stamp_inner,
        source_page=source_page,
        target_pages=target_pages,
        x=x,
        y=y,
        scale=scale,
        opacity=opacity,
        layer=layer,
    )
    doc = Document.__new__(Document)
    doc._inner = new_inner
    doc._closed = False
    doc._raw_bytes = self._raw_bytes
    return doc


Document.stamp = _stamp  # type: ignore[method-assign]

File: crates/paperjam-model/src/stamp.rs
/// Whether the stamp goes over or under existing content.
#[derive(Debug, Clone)]
pub enum StampLayer {
    Over,
    Under,
}

impl StampLayer {
    #[allow(clippy::should_implement_trait)]
    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "under" | "underlay" => Self::Under,
            _ => Self::Over,
        }
    }
}

/// Options for page stamping/overlay.
pub struct StampOptions {
    /// 1-based page number in the stamp document to use as the stamp.
    pub source_page: u32,
    /// Target pages in the document. None = all pages.
    pub target_pages: Option<Vec<u32>>,
    /// X offset in points.
    pub x: f64,
    /// Y offset in points.
    pub y: f64,
    /// Scale factor (default 1.0).
    pub scale: f64,
    /// Opacity (0.0-1.0, default 1.0).
    pub opacity: f64,
    /// Whether stamp goes over or under existing content.
    pub layer: StampLayer,
}

impl Default for StampOptions {
    fn default() -> Self {
        Self {
            source_page: 1,
            target_pages: None,
            x: 0.0,
            y: 0.0,
            scale: 1.0,
            opacity: 1.0,
            layer: StampLayer::Over,
        }
    }
}
